sim/core/attack.go

```diff
@@ -245,76 +248,101 @@ func (ama *ActiveMeleeAbility) Attack(sim *Simulation) bool {
+		}
 		// Not sure MH/OH Matters for an attack
 		ama.Character.OnMeleeAttack(sim, ama.Target, hit, ama, false)
-		return true
 	}
 
 	c := ama.Character
-
-	if ama.DirectDamageInput.FlatDamageBonus > 0 || ama.DirectDamageInput.MinBaseDamage > 0 {
-		// Do a 'direct damage' if ability has it
-		dmg := ama.DirectDamageInput.MinBaseDamage + (ama.DirectDamageInput.MaxBaseDamage-ama.DirectDamageInput.MinBaseDamage)*sim.RandomFloat("melee direct damage") + ama.DirectDamageInput.FlatDamageBonus
-		c.Metrics.TotalDamage += dmg
-		if sim.Log != nil {
-			sim.Log("%s for %0.1f", ama.Name, dmg)
-		}
-	}
-
-	// Only calculate attack if there is a weapon swing involved.
-	if ama.WeaponDamageInput.MainHand == 0 && ama.WeaponDamageInput.Offhand == 0 {
-		return true
-	}
-
 	skill := 350.0
 	level := float64(ama.Target.Level)
 	critReduction := (level - 70*0.01) + 0.018
 	critChance := c.stats[stats.MeleeCrit]/(MeleeCritRatingPerCritChance*100) + (skill - (level*5)*0.002) - critReduction
-	roll := sim.RandomFloat("weapon swing")
+
+	if ama.DirectDamageInput.FlatDamageBonus > 0 || ama.DirectDamageInput.MinBaseDamage > 0 {
+		ama.applyFlatDamage(sim, critChance)
+	}
+
 	if ama.WeaponDamageInput.MainHand > 0 {
-		weapon := c.Equip[proto.ItemSlot_ItemSlotMainHand]
-		speed := ama.NormalizeWeaponSpeed
-		if speed == 0 {
-			speed = weapon.SwingSpeed
-		}
-		dmgMult := ama.WeaponDamageInput.MainHand
-		if roll < critChance {
-			dmgMult *= 2.0
-		}
-		dmg := weapon.WeaponDamageMin + (weapon.WeaponDamageMax-weapon.WeaponDamageMin)*sim.RandomFloat("auto attack")
-		dmg += (speed * c.stats[stats.AttackPower]) / MeleeAttackRatingPerDamage
-		dmg *= dmgMult
-		dmg += ama.WeaponDamageInput.MainHandFlat
-		c.Metrics.TotalDamage += dmg
-		if sim.Log != nil {
-			sim.Log("%s mainhand for %0.1f", ama.Name, dmg)
-		}
-		c.OnMeleeAttack(sim, ama.Target, ama.Result, ama, false)
+		ama.applySwingDamage(sim, proto.ItemSlot_ItemSlotMainHand, ama.WeaponDamageInput.MainHand, critChance)
 	}
 
 	if weapon := c.Equip[proto.ItemSlot_ItemSlotOffHand]; ama.WeaponDamageInput.Offhand > 0 && weapon.ID > 0 { // only attack if we have it
-		speed := ama.NormalizeWeaponSpeed
-		if speed == 0 {
-			speed = weapon.SwingSpeed
-		}
-		dmgMult := ama.WeaponDamageInput.Offhand * 0.5
-		if roll < critChance {
-			dmgMult *= 2.0
-		}
-		dmg := weapon.WeaponDamageMin + (weapon.WeaponDamageMax-weapon.WeaponDamageMin)*sim.RandomFloat("auto attack")
-		dmg += (speed * c.stats[stats.AttackPower]) / MeleeAttackRatingPerDamage
-		dmg *= dmgMult
-		dmg += ama.WeaponDamageInput.OffhandFlat
-		c.Metrics.TotalDamage += dmg
-		if sim.Log != nil {
-			sim.Log("%s offhand for %0.1f", ama.Name, dmg)
-		}
-		c.OnMeleeAttack(sim, ama.Target, ama.Result, ama, true)
+		ama.applySwingDamage(sim, proto.ItemSlot_ItemSlotOffHand, ama.WeaponDamageInput.Offhand, critChance)
 	}
 
 	return true
 }
 
+func (ama *ActiveMeleeAbility) applySwingDamage(sim *Simulation, slot proto.ItemSlot, dmgMult, critChance float64) {
```

Comment:
Really don't like the name 'ama', its hard to remember what it means while reading the function. If you don't want to write out the full object name, you could use 'ability' or 'meleeAbility'.